File: examples/hello-receipt-verify/verify.py
"""Verify every exported receipt using separately selected trusted signers."""
import argparse
import json
from pathlib import Path
from chio.invariants import verify_receipt_with_trusted_signers


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('receipts', nargs='?', type=Path, default=Path('fixtures/minimal-evidence/receipts.ndjson'))
    parser.add_argument('--trusted-signers', type=Path, default=Path('trusted-signers.json'))
    args = parser.parse_args()
    signers = json.loads(args.trusted_signers.read_text())
    if not isinstance(signers, list) or not signers or not all(isinstance(key, str) for key in signers):
        parser.error('trusted-signers must be a nonempty JSON array of public keys')
    records = [json.loads(line) for line in args.receipts.read_text().splitlines() if line.strip()]
    if not records:
        parser.error('receipt file is empty')
    results = [verify_receipt_with_trusted_signers(record.get('receipt', record), signers) for record in records]
    print(json.dumps(results, indent=2))
    return 0 if all(result['ok'] for result in results) else 1


if __name__ == '__main__':
    raise SystemExit(main())

File: examples/hello-receipt-verify/verify.go
package main

import (
	"bufio"
	"encoding/json"
	"flag"
	"fmt"
	"github.com/backbay-labs/chio/sdks/go/chio-go/invariants"
	"os"
)

func run() error {
	trust := flag.String("trusted-signers", "trusted-signers.json", "JSON array of independently selected trusted public keys")
	flag.Parse()
	receiptPath := "fixtures/minimal-evidence/receipts.ndjson"
	if flag.NArg() > 0 {
		receiptPath = flag.Arg(0)
	}
	keys, err := os.ReadFile(*trust)
	if err != nil {
		return err
	}
	var signers []string
	if err := json.Unmarshal(keys, &signers); err != nil {
		return err
	}
	if len(signers) == 0 {
		return fmt.Errorf("trusted signer set is empty")
	}
	f, err := os.Open(receiptPath)
	if err != nil {
		return err
	}
	defer f.Close()
	scanner := bufio.NewScanner(f)
	scanner.Buffer(make([]byte, 4096), 4*1024*1024)
	count := 0
	valid := true
	for scanner.Scan() {
		if len(scanner.Bytes()) == 0 {
			continue
		}
		var record map[string]any
		if err := json.Unmarshal(scanner.Bytes(), &record); err != nil {
			return err
		}
		receipt, ok := record["receipt"].(map[string]any)
		if !ok {
			receipt = record
		}
		result, err := invariants.VerifyReceiptWithTrustedSigners(receipt, signers)
		if err != nil {
			return err
		}
		out, err := json.MarshalIndent(result, "", "  ")
		if err != nil {
			return err
		}
		fmt.Println(string(out))
		valid = valid && result.Ok
		count++
	}
	if err := scanner.Err(); err != nil {
		return err
	}
	if count == 0 || !valid {
		return fmt.Errorf("receipt verification failed")
	}
	return nil
}
func main() {
	if err := run(); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
}
